xmrstak/backend/amd/autoAdjust.hpp

```diff
@@ -97,34 +95,46 @@ class autoAdjust
 		int i = 0;
 		for(auto& ctx : devVec)
 		{
-			/* 1000 is a magic selected limit, the reason is that more than 2GiB memory
-			 * sowing down the memory performance because of TLB cache misses
-			 */
-			size_t maxThreads = 1000u;
-			if(ctx.name.compare("gfx901") == 0)
-			{
-				/* Increase the number of threads for AMD VEGA gpus.
-				 * Limit the number of threads based on the issue: https://github.com/fireice-uk/xmr-stak/issues/5#issuecomment-339425089
-				 * to avoid out of memory errors
-				 */
-				maxThreads = 2024u;
-			}
-
-			// keep 128MiB memory free (value is randomly chosen)
-			size_t availableMem = ctx.freeMem - (128u * byteToMiB);
-			// 224byte extra memory is used per thread for meta data
+			/*
+			  GCN has a hard architectural limit of 64 CUs max (Invalid for other archs though).
+			  Reserve 1 CU for the system. 
+			*/
+			size_t maxComputeUnitsAvailable = std::max(std::min(ctx.computeUnits, 64) - 1, 1);
+			//Each CU is 64 threads executed in an SIMT fashion:
+			size_t maxThreadsAvailable = maxComputeUnitsAvailable << 6;
```

Comment:
Please not use magic bit shift, try to keep the code human readable. `X << 6` is equal to `X * 64` the compiler will replace it with a bit shift.

Reply:
The ">> 7"'s have a purpose - they round down to units of 64 threads for the dual thread optimization. The << 6's though have been changed back to * 64. I reversed the multiply and divide by one megabyte change.